addons/pos_online_payment/controllers/payment_portal.py

```diff
@@ -0,0 +1,299 @@
+
+    @staticmethod
+    def _is_valid_amount(amount, currency):
+        return isinstance(amount, float) and tools.float_compare(amount, 0.0, precision_rounding=currency.rounding) > 0
+
+    @staticmethod
+    def _get_allowed_providers_sudo(request, pos_order_sudo, partner_id, amount_to_pay):
+        payment_method = pos_order_sudo.online_payment_method_id
+        if not payment_method:
+            raise UserError(_("There is no online payment method configured for this Point of Sale order."))
+        compatible_providers_sudo = request.env['payment.provider'].sudo()._get_compatible_providers(
+            pos_order_sudo.company_id.id, partner_id, amount_to_pay, currency_id=pos_order_sudo.currency_id.id)
+        # In sudo mode to read the fields of providers and partner (if not logged in)
+        # Return the payment providers configured in the pos.payment.method that are compatible for the payment API
+        return compatible_providers_sudo & payment_method._get_online_payment_providers(pos_order_sudo.config_id.id, error_if_invalid=False)
+
+    @staticmethod
+    def _get_exit_route_arg(exit_route):
+        return f'&exit_route={urllib.parse.quote(exit_route)}' if exit_route else ''
+
+    @staticmethod
+    def _get_pay_route(pos_order_id, access_token, exit_route=None):
+        exit_route_arg = PaymentPortal._get_exit_route_arg(exit_route)
+        return f'/pos/pay/{pos_order_id}?access_token={access_token}{exit_route_arg}'
+
+    @staticmethod
+    def _get_landing_route(pos_order_id, access_token, exit_route_arg=None, tx_id=None):
+        tx_id_arg = f'&tx_id={tx_id}' if tx_id else ''
+        return f'/pos/pay/confirmation/{pos_order_id}?access_token={access_token}{exit_route_arg}{tx_id_arg}'
+
+    @http.route('/pos/pay/<int:pos_order_id>', type='http', methods=['GET'], auth='public', website=True, sitemap=False)
+    def pos_order_pay(self, pos_order_id, access_token=None, exit_route=None):
+        """ Behaves like payment.PaymentPortal.payment_pay but for POS online payment.
+
+        :param str pos_order_id: The POS order to pay, as a `pos.order` id
+        :param str access_token: The access token used to verify the user
+        :param str exit_route: The URL to open to leave the POS online payment flow
+
+        :return: The rendered payment form
+        :rtype: str
+        :raise: AccessError if the provided order or access token is invalid
+        :raise: ValidationError if data on the server prevents the payment
+        """
+        pos_order_sudo = self._check_order_access(pos_order_id, access_token)
+        self._ensure_session_open(pos_order_sudo)
+
+        user_sudo = request.env.user
+        logged_in = not user_sudo._is_public()
+        partner_sudo = self._get_partner_sudo(user_sudo)
+        if not partner_sudo:
+            return self._redirect_login(request)
+
+        kwargs = {
+            'pos_order_id': pos_order_sudo.id,
+        }
```

Comment:
Maybe today, but tomorrow if you want to add something, you just need to add it to this dict.
I try to use as little copy/paste I can, because duplicating stuff usually ends up with incoherent code, particularly in a community project as Odoo where everyone doesn't take the same care when changing something.